File: packages/sdk/src/discovery.ts
export interface DiscoveryDocument {
  issuer: string;
  authorization_endpoint: string;
  token_endpoint: string;
  userinfo_endpoint: string;
  jwks_uri: string;
  end_session_endpoint?: string;
}

export async function fetchDiscoveryDocument(issuer: string): Promise<DiscoveryDocument> {
  const base = issuer.replace(/\/$/, "");
  const res = await fetch(`${base}/.well-known/openid-configuration`);
  if (!res.ok) {
    throw new Error(`failed to fetch OIDC discovery document from ${issuer} (${res.status})`);
  }
  return (await res.json()) as DiscoveryDocument;
}

File: packages/sdk-go/discovery.go
package neko

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

type DiscoveryDocument struct {
	Issuer                string `json:"issuer"`
	AuthorizationEndpoint string `json:"authorization_endpoint"`
	TokenEndpoint         string `json:"token_endpoint"`
	UserinfoEndpoint      string `json:"userinfo_endpoint"`
	JWKSURI               string `json:"jwks_uri"`
}

func FetchDiscoveryDocument(issuer string) (*DiscoveryDocument, error) {
	base := strings.TrimRight(issuer, "/")
	res, err := http.Get(base + "/.well-known/openid-configuration")
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch OIDC discovery document from %s (%d)", issuer, res.StatusCode)
	}
	var doc DiscoveryDocument
	if err := json.NewDecoder(res.Body).Decode(&doc); err != nil {
		return nil, err
	}
	return &doc, nil
}
